### DISCUSSION TITLE
Set application icon

### DISCUSSION THREAD

**Discussion by KiraResari** [hecrj/iced#1281] [Q&A] [answered]
I am now trying to set an application icon to my iced application. I've tried this as follows:
```
pub fn main() -> iced::Result {
    let mut settings = Settings::default();
    settings.window.icon =  Icon::from_rgba(
        include_bytes!(
            "../resources/images/Kanji Tree R Icon.ico"
        ).to_vec(),
        512,
        512
    ).ok();
    kanji_tree::KanjiTreeApp::run(settings)
}
```

However, in that case, `from_rgba` returns the following error:

```
DimensionsMismatch
* width: 512
* height: 512
* pixel_count: 34359
```

I've also tried this with icons of different sizes and file types, and sometimes it returns an error like the following instead (for a 32x32 .ico):

```
InvalidData
* byte_count: 4286
* [raw]
** variant0:
*** byte_count: 4286
** variant1:
*** width: 385
*** height: 4286
*** pixel_count: 1653562408960
```

I haven't been able to find a way to get this to work. Can someone please help me with that?

**Comment by BlackHoleFox** [comment, marked as answer]
You need to decode the image data into raw bytes first.

Here's an example which uses the `image` crate to do that. Replace the constants with the properties/location/etc of your actual icon.
```rust
static ICON: &[u8] = include_bytes!("../icon.png");
const ICON_HEIGHT: u32 = 64;
const ICON_WIDTH: u32 = 64;

fn main() {
    let image = image::load_from_memory(ICON).unwrap();
    let icon = iced::window::Icon::from_rgba(image.as_bytes().to_vec(), ICON_HEIGHT, ICON_WIDTH).unwrap();
    
    let settings = iced::settings::Settings {
        window: iced::window::Settings {
            icon: Some(icon), 
            ..Default::default()
        },
        ..Default::default()
    };

    YourApplication::run(settings)
    ...
}

**Reply by KiraResari** [reply]
Cool, thanks! Is there a way to get .ico icons to work too?

**Reply by BlackHoleFox** [reply]
The same thing should work. You just need to enable the `ico` feature of `image`.

If you want to make the icon show up as the executable's icon itself though (assuming this is Windows), you'll want to use the `winres` crate.

**Reply by KiraResari** [reply]
I already managed to make the icon show up in the compiled exe.

However, using the .ico for the app icon isn't working yet.

So, I did enable the ico feature of image like this:
```
image = {version = "0.24.1", features = ["ico"] }
```

...and then I tried calling it like this:
```
static ICON: &[u8]
     = include_bytes!("../resources/images/Kanji Tree R Icon.ico");
const ICON_HEIGHT: u32 = 512;
const ICON_WIDTH: u32 = 512;

pub fn main() -> iced::Result {
    let settings = build_settings();
    kanji_tree::KanjiTreeApp::run(settings)
}

fn build_settings() -> Settings<()> {
    let image
        = image::load_from_memory(ICON).unwrap();
    let icon = Icon::from_rgba(
        image.as_bytes().to_vec(),
        ICON_HEIGHT,
        ICON_WIDTH
    ).unwrap();
    Settings {
        window: iced::window::Settings {
            icon: Some(icon), 
            ..Default::default()
        },
        ..Default::default()
    }
}
```

However, then I get this error:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Decoding(DecodingError { format: Exact(Ico), underlying: Some(ImageEntryDimensionMismatch { format: Png, entry: (256, 256), image: (512, 512) }) })', src\main.rs:17:41
```

I checked, and the ico dimensions are 512x512. However, I notice that the `format: Png` is wrong here, and I tried reading the documentation of image, but wasn't able to figure out what I need to change to make it use .ico instead. Can you maybe elaborate?

Also, if I use the same code and specify a .png instead of an .ico (I have both versions lying side by side in my resources folder) it works just fine.

**Reply by BlackHoleFox** [reply]
Seems like the image data embedded in the `.ico` file is too large? [Wikipedia says](https://en.wikipedia.org/wiki/ICO_(file_format)) the format supports up to 256x256 and the [image source code](https://github.com/image-rs/image/blob/5893be0a1fa2da8a36de08f63d55f9eb95b2cfc4/src/codecs/ico/decoder.rs#L169-L170) also seems to agree. Meanwhile, the actual png inside is 512x512 according to that error from [here](https://github.com/image-rs/image/blob/5893be0a1fa2da8a36de08f63d55f9eb95b2cfc4/src/codecs/ico/decoder.rs#L313-L323).

**Reply by KiraResari** [reply]
Ah, looks like that was it! Thanks for pointing that out! It now changed it to 256x256, and it works just fine.

The interesting thing here is, however, that the `build.rs` was able to handle the 512x512 ico file just fine. I suppose they use different approaches for handling ico files.

**Reply by BlackHoleFox** [reply]
I'd not be shocked at all if `rc.exe` (the thing `winres` invokes) just silently shrunk larger icons for more broad compatibility.

**Comment by dassio** [comment]
i am using iced 0.4.2 on windows 10 with the accepted answer code, but still get the `DimensionsMismatch` error, anyone else experiencing this ?
